Use conditions for volume (instead of Phase, ErrorCode and ErrorMessage)

slaperche-scality opened the pull request:
**Component**:

storage, ui, tests

**Context**: 

Update the Volume CRD to use a list of `Conditions` instead of the `Phase/ErrorCode/ErrorMessage` triptych

**Summary**:

- update the CRD
- update the operator code with a compat layer
- update the E2E tests with a compat layer
- update the UI code with a compat layer
- add some unit tests for the compat layer

**Acceptance criteria**: 

Everything should work as before, no visible change (except the Volume schema of course)

---

Closes: #1507

slaperche-scality committed:
storage-operator/volume: add `Conditions` fields

This commit introduces a new field in `VolumeStatus`: `Conditions`. This
field contains a list of `VolumeCondition` and supersedes the fields
`Phase`, `ErrorCode` and `ErrorMessage`.

Using a list of conditions allows for more flexibility (we can add new
conditions without impacting the CRD) and is the canonical way to
represent status in the Kubernetes world.

Refs: #1507
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator: adapt code to use conditions

This is the first step toward using the conditions.
We won't harvest the benefits of using conditions yet, this commit is
mainly to introduce the schema changes while keeping the old logic of a
"global" status.

For now we are using a single condition, Ready, which can be used to
compute a "global" status and allows us to keep most of the code
unchanged.

The mapping is as follow:

    | `status.phase` | `Conditions[Ready]`                   |
    | -------------- | ------------------------------------- |
    | `Unknown`      | missing                               |
    | `Failed`       | `False`                               |
    | `Pending`      | `Unknown` + `Reason` ==  `Pending`    |
    | `Available`    | `True`                                |
    | `Terminating`  | `Unknown` + `Reason` == `Terminating` |

Refs: #1507
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests/volume: update tests to use conditions

Like for the operator code, we want to keep believing that we have a
global status with a single error code/message, thus the `features` are
unchanged: we simply add a thin layer of conversion in the `steps` to
convert the conditions into the old paradigm.

Refs: #1507
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
ui/volume: adapt code to use conditions

Still striving to minimize the changes, we only fix the creation of the
Volume object in the UI (in `NodeInformation.js`) by using a function to
compute a global status.
This allows us to keep most of the code unchanged (no need to rework the
table or any visible component), which reduces the risk of regression.

Closes: #1507
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality

slaperche-scality committed:
storage-operator/tests: add unit test for the compat layer

Add some basic unit tests for the compat helper that deals with
conditions.

Refs: #1507
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
ui/tests: add unit tests for condition compat layer

Add some basic unit tests to cover the helpers that ensure the
conversion between conditions and old global status paradigm.

Refs: #1507
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

charlesprost commented on storage-operator/deploy/crds/storage_v1alpha1_volume_crd.yaml:
```diff
@@ -78,28 +78,42 @@ spec:
+              items:
+                properties:
+                  lastTransitionTime:
+                    description: Last time the condition transit from one status to
```
```suggestion
                    description: Last time the condition transited from one status to
```

slaperche-scality replied:
This file is auto generated, but I'll fix the comment from which this is generated.
Done

charlesprost commented on storage-operator/deploy/crds/storage_v1alpha1_volume_crd.yaml:
```diff
@@ -78,28 +78,42 @@ spec:
+                properties:
+                  lastTransitionTime:
+                    description: Last time the condition transit from one status to
+                      another (optional)
```
```suggestion
                      another (optional).
```

slaperche-scality replied:
ditto and done.

charlesprost commented on storage-operator/pkg/apis/storage/v1alpha1/volume_types.go:
```diff
@@ -63,44 +63,72 @@ type PersistentVolumeTemplateSpec struct {
 type VolumePhase string
 
-// "Enum" representing the phase/status of a volume.
+// TODO: kept for temporary compatibility, to be removed.
```
Temporary? What would condition the end of this temporary time?

slaperche-scality replied:
When we will have fine-grained conditions representing each steps, the concept of global phase may become irrelevant/useless and thus it could be removed (some k8s objects only have conditions).
I'll try to open a new issue about this, containing the more completed design I first proposed for this one (but which I then simplified due to the time constraint)

charlesprost replied:
OK thx for the explanation.

charlesprost commented on storage-operator/pkg/apis/storage/v1alpha1/volume_types.go:
```diff
@@ -63,44 +63,72 @@ type PersistentVolumeTemplateSpec struct {
 
-// "Enum" representing the phase/status of a volume.
+// TODO: kept for temporary compatibility, to be removed.
+// "Enum" representing the phase of a volume.
```
So we still have a phase for a volume but it's not directly in the CRD but it comes from the computation of all conditions?

charlesprost replied:
Oh I've just discovered the `computePhase` later, so I guess that's it.

slaperche-scality replied:
Yeah, we still have the concept of a global "Phase" because lot of logic relies on that.

Not sure we will keep it when we make the real move to conditions, we will see.
But for now we have a computed one to ease the transition.

charlesprost commented on storage-operator/pkg/apis/storage/v1alpha1/volume_types.go:
```diff
@@ -63,44 +63,72 @@ type PersistentVolumeTemplateSpec struct {
-type VolumeErrorCode string
+type ConditionReason string
 
+// TODO: replace those by more fine-grained one.
```
```suggestion
// TODO: replace those by more fine-grained ones.
```

slaperche-scality replied:
Done

charlesprost commented on storage-operator/pkg/apis/storage/v1alpha1/volume_types.go:
```diff
@@ -63,44 +63,72 @@ type PersistentVolumeTemplateSpec struct {
+	Status ConditionStatus `json:"status"`
+	// Last time the condition was updated (optional).
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
+	// Last time the condition transit from one status to another (optional)
```
```suggestion
	// Last time the condition transited from one status to another (optional).
```

slaperche-scality replied:
Done

charlesprost reviewed (commented):
Very minor stuff to fix (or not).
Not sure I'm able to review the Go syntax but the logic seems clear (or I'm just tired maybe ^^).
Didn't test it but if our tests are green, that should be fine.
Good job.

NicolasT commented on storage-operator/pkg/apis/storage/v1alpha1/volume_types.go:
```diff
@@ -63,44 +63,72 @@ type PersistentVolumeTemplateSpec struct {
+	VolumeReady VolumeConditionType = "Ready"
 )
 
+type ConditionStatus string
```
Can't we reuse the type from `k8s.io/pkg/apis/core` (or whatever the right import path is)?

slaperche-scality replied:
Ho yes, you're right, let's use `corev1.ConditionStatus`

NicolasT commented on storage-operator/pkg/apis/storage/v1alpha1/volume_types.go:
```diff
@@ -63,44 +63,72 @@ type PersistentVolumeTemplateSpec struct {
+type ConditionStatus string
+
+const (
+	ConditionTrue    ConditionStatus = "True"
```
(and then these values as well)

slaperche-scality replied:
Done

NicolasT reviewed (approved):
Somewhat of a cursory review, but LGTM!

slaperche-scality commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/7f83e677de00989609abd588b180ec04c9eb53d5) did not succeed in branch improvement/1507-use-conditions-for-volume.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
